crates/biome_js_analyze/src/lint/nursery/no_undeclared_env_vars.rs

```diff
@@ -0,0 +1,247 @@
+        let env_var = env_var_text.text();
+
+        // Check if this env var is allowed by default patterns or user options
+        if is_env_var_allowed_by_defaults(env_var)
+            || is_env_var_allowed_by_options(env_var, options)
+        {
+            return None;
+        }
+
+        // Check if declared in turbo.json
+        if ctx.is_env_var_declared(env_var) {
+            return None;
+        }
+
+        Some(EnvVarAccess {
+            env_var_name: env_var_text,
+        })
+    }
+
+    fn diagnostic(ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        let node = ctx.query();
+        let env_var = state.env_var_name.text();
+
+        Some(RuleDiagnostic::new(
+            rule_category!(),
+            node.range(),
+            markup! {
+                "The environment variable "<Emphasis>{env_var}</Emphasis>" is not listed as a dependency in turbo.json. Add this environment variable to 'globalEnv', 'globalPassThroughEnv', or a task's 'env'/'passThroughEnv' array in your turbo.json(c) configuration to ensure correct caching behavior in Turborepo."
+            },
+        ))
+    }
+}
+
+/// Checks if the object is the global `process` identifier (not shadowed by a local binding)
+fn is_global_process_object(
+    expr: &AnyJsExpression,
+    model: &biome_js_semantic::SemanticModel,
+) -> bool {
+    let Some((reference, name)) = global_identifier(expr) else {
+        return false;
+    };
+
+    // Check that it's named "process" and not bound to a local variable
+    name.text() == "process" && model.binding(&reference).is_none()
+}
+
+/// Checks if the object is `import.meta`
+fn is_import_meta_object(expr: &AnyJsExpression) -> bool {
+    expr.as_js_import_meta_expression().is_some()
+}
+
+/// Default allowed environment variables that are commonly used
+const DEFAULT_ALLOWED_ENV_VARS: &[&str] = &[
+    "NODE_ENV", "CI", "TZ", "PATH", "HOME", "USER", "SHELL", "PWD",
+];
+
+/// Pre-compiled regex patterns for framework and CI provider specific env vars (e.g., NEXT_PUBLIC_*, VITE_*, etc.)
+/// Using LazyLock ensures these are compiled only once and reused across all rule invocations.
+static NEXT_PUBLIC_PATTERN: LazyLock<Regex> =
+    LazyLock::new(|| Regex::new(r"^NEXT_PUBLIC_.*$").unwrap());
+static VITE_PATTERN: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^VITE_.*$").unwrap());
+static REACT_APP_PATTERN: LazyLock<Regex> =
+    LazyLock::new(|| Regex::new(r"^REACT_APP_.*$").unwrap());
+static VUE_APP_PATTERN: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^VUE_APP_.*$").unwrap());
+static NUXT_PATTERN: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^NUXT_.*$").unwrap());
+static GATSBY_PATTERN: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^GATSBY_.*$").unwrap());
+static EXPO_PUBLIC_PATTERN: LazyLock<Regex> =
+    LazyLock::new(|| Regex::new(r"^EXPO_PUBLIC_.*$").unwrap());
+static VERCEL_PATTERN: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^VERCEL(_.*)?$").unwrap());
```

Comment:
Its a bit unusual for us to use regex for these. We tend to avoid it because of performance.

Reply:
I saw this pattern in other files while I was looking for regex handling patterns for the other regex part of this PR and put it down in my notes without even thinking about just being normal and doing The Simple Thing (`starts-with`) for this part. 🤦